### ISSUE TITLE
Data Augmentation in Image Classification

### ISSUE THREAD

**Issue by vsvanshi** [ludwig-ai/ludwig#490]
Hi,
How can we enable Data augmentation in Image classification?

**Comment by w4nderlust** [comment]
There is no data augmentation feature at the moment.
We definitely want to add it in the future.
Contributions are welcome.

**Comment by w4nderlust** [comment]
@manesioz the example you provide is useful. The problem with it is that it looks like it assumes that the inputs to your model are only images and that the outputs to your model are only categories judging from this snippet: https://github.com/tensorflow/tensorflow/blob/r1.14/tensorflow/python/keras/preprocessing/image.py#L376-L403
In Ludwig that's not the case, so if someone wants to add the same functionality in Ludwig, one would have to deal with other input features and potentially multiple output features.

**Comment by manesioz** [comment]
@w4nderlust thanks for the advice, I agree that this feature should be generally available for different inputs/outputs. However, considering how most data augmentation techniques are highly domain-specific (image, text, timeseries, etc), I was wondering if we should initially approach this feature as being configurable depending on the input data type (text, image, numerical, timeseries, etc)?   

Some brief [literature review](https://arxiv.org/pdf/1702.05538.pdf) indicates that some domain-agnostic techniques exist, but I am not sure of their accuracy/performance/general acceptance within the ML community. 

Any advice would be much appreciated, thanks!

**Comment by w4nderlust** [comment]
@manesioz  sorry I did not mean that the data augmentation should work for all feature types, what I meant is that the example works with only one feature input, in Ludwig there could be several, so for instance if you apply 4 transformations to your image input feature, all the other input features have to be replicated 4 more times, which the example does not consider.

**Comment by carlogrisetti** [comment]
Just for a reference, I am very proficiently using ImgAug (https://github.com/aleju/imgaug) for image augmentation.
Has *lots* of augmentation possibilities, and correctly maps points\vectors\shapes\segmentations\etc on the destination image\array.
It takes account of any chained modifications on the image (multiple transformations do not require multiple adaptations of the other features, but are passed in one unique transformation sequence) and has many many transformations built-in (see the project homepage for a visual demo)

I agree with the fact that augmentation should not be domain agnostic... In the last 2 or 3 projects I used data augmentation on (always and only talking about images) I had to perform very different image augmentations based on the dataset I'm working on.
Some had to never have flipped or distorted images, but were highly improved by noise\sharpening\blurring\zoom\crop\etc, while others needed not to have any form of noise\sharpening\blurring and were highly improved by flipping\distorting\rotating\etc
Just think about a specific object detection in an industrial\machining environment (flipping would be a huge issue on which the piece would have to be discarded while noise is not) and about a production\textile environment (flipping a piece of fabric does absolutely nothing, distortions are already present, but noise and color noise would be disrupting in a quality control scenario)

**Comment by w4nderlust** [comment]
I think it could be fine to have several options (like tokenizers for text) which could be domain specific as long as the user can decide what to use fro the model definition :)

**Comment by andife** [comment]
Hi, are there any considerations / developments to introduce image augmentation for example image classification problems? Has anyone already extended ludwig in this regard?

**Comment by w4nderlust** [comment]
@andife unfortunately no updates. There will soon be updates on using pretrained image encoders and having more architectures available though.
This is still a feature we want to add, but has not been prioritized yet.
Still very open for community contributions about it!

**Comment by andife** [comment]
@w4nderlust : Yes definitly an option. I'll look at the code.  Maybe it's not too much to for example to add torchvision.transforms, kornia or albumentations there.

**Comment by jimthompson5802** [comment]
Closed by #2925
